packages/Operators/src/PointCloud/DTK_NodeToNodeOperator_impl.hpp

```diff
@@ -0,0 +1,252 @@
+    Teuchos::ArrayRCP<double> source_centers( DIM*local_num_src);
+    Teuchos::ArrayRCP<GO> source_support_ids( local_num_src );
+    Teuchos::Array<SupportId> source_node_supports;
+    EntityIterator domain_begin = domain_iterator.begin();
+    EntityIterator domain_end = domain_iterator.end();
+    int entity_counter = 0;
+    for ( EntityIterator domain_entity = domain_begin;
+	  domain_entity != domain_end;
+	  ++domain_entity, ++entity_counter )
+    {
+	domain_space->shapeFunction()->entitySupportIds(
+	    *domain_entity, source_node_supports );
+	DTK_CHECK( 1 == source_node_supports.size() );
+	source_support_ids[entity_counter] = source_node_supports[0];
+	domain_space->localMap()->centroid(
+	    *domain_entity, source_centers(DIM*entity_counter,DIM) );
+    }
+
+    // Extract the target nodes and their ids.
+    EntityIterator range_iterator;
+    if ( nonnull_range )
+    {
+	LocalEntityPredicate local_predicate(
+	    range_space->entitySet()->communicator()->getRank() );	
+	PredicateFunction range_predicate =
+	    PredicateComposition::And(
+		range_space->selectFunction(), local_predicate.getFunction() );
+	range_iterator =
+	    range_space->entitySet()->entityIterator( 0, range_predicate );
+    } 
+    int local_num_tgt = range_iterator.size();
+    Teuchos::ArrayRCP<double> target_centers( DIM*local_num_tgt );
+    Teuchos::ArrayRCP<GO> target_support_ids( local_num_tgt );
+    Teuchos::Array<SupportId> target_node_supports;
+    EntityIterator range_begin = range_iterator.begin();
+    EntityIterator range_end = range_iterator.end();
+    entity_counter = 0;
+    for ( EntityIterator range_entity = range_begin;
+	  range_entity != range_end;
+	  ++range_entity, ++entity_counter )
+    {
+	range_space->shapeFunction()->entitySupportIds(
+	    *range_entity, target_node_supports );
+	DTK_CHECK( 1 == target_node_supports.size() );
+	target_support_ids[entity_counter] = target_node_supports[0];
+	range_space->localMap()->centroid(
+	    *range_entity, target_centers(DIM*entity_counter,DIM) );
+    }
+
+    // Calculate an approximate neighborhood distance for the local target
+    // centers. 
+    double target_proximity = 0.0;
+
+    // Get the local bounding box.
+    Teuchos::Tuple<double,6> local_box;
+    range_space->entitySet()->localBoundingBox( local_box );
+
+    // Calculate the largest span of the cardinal directions.
+    target_proximity = local_box[3] - local_box[0];
+    for ( int d = 1; d < DIM; ++d )
+    {
+        target_proximity = std::max( target_proximity,
+                                     local_box[d+3] - local_box[d] );
+    }
+
+    // Take the proximity to be 0.001 of the largest distance.
+    target_proximity *= 1.0e-3;
+    
+    // Gather the source centers that are in the proximity of the target
+    // centers on this proc.
+    Teuchos::Array<double> dist_sources;
+    CenterDistributor<DIM> distributor( 
+	comm, source_centers(), target_centers(), target_proximity, dist_sources );
```

Comment:
Remind me where the kd-tree is built. It sounds silly to me that we use the proximity if the tree already exists.
In that case I would just do a ``nnSearch`` for k=1.